docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,112 @@
+
+* Promotion and demotion of RBD image.
+
+## Below provides the outline of how this can be implemented in ceph-csi
+
+* A storage-class Need to have a parameter to tell that the image needs to be
+  mirrored.
+* The PVC name and namespace need to be passed to the CSI driver as part of
+  volume create request (rbd image and omap keys/value will be derived from
+  these names if its mirrored image.
+* The RBD mirror daemon required setup has to be done prior enabling the image
+  mirroring from ceph-csi.
+* Promotion and Demotion of RBD image.
+
+### Storage-class parameter
+
+we need to add a new parameter to the storage-class to indicate the image need
+to be mirrored upon creation, based on this parameter ceph-csi will decide few
+things like (information need to store and omap for image identification,
+finding rbd image based on PVC name and namespace name in the remote cluster)
+
+The value for the mirroring will be  either one of these "disabled" (default),
+"journal", or "snapshot"
+
+* "disabled" - (default) no need to mirror the rbd image
+* "journal"  - journal based mirroring. The mounter should be rnd-nbd as the
+  krbd doesnot support it yet.
+* "snapshot" - snapshot based mirroring. It should work with existing krbd
+  (need ceph octopus+)
+
+More info on rbd mirroring at [1].
+
+### Pass PVC name and namespace name in CreateVolume
+
+If the app and the PVC created in the local cluster as part of recovery the app
+and the PVC will be recreated in the remote cluster, the data present in the
+PVC in remote  cluster should be the same in the local cluster.
+
+As the PVC name and the namespace name should be sent in volume create request,
+cephcsi will check that the rbd image is already present cluster, we cannot
+check the omap data first as omap data maintained by ceph-csi is not mirrored
+to the remote cluster. This is the challenge ceph-csi need to handle)If the
+image is not present(means its a local cluster) we need to generate the
+required omap data and store its information and create the rbd image, the CSI
+will need to enable mirroring on the newly created RBD image via `rbd mirror
+image enable image-spec`.
+
+If it is a remote cluster the image will be already created and mirrored by the
+local cluster, we need to check for the image first(if it is present we need to
+generate the omap data and return success to the create volume (as its a remote
+cluster we don't need to enable mirroring here.
+
+Incase of journal based mirroring the image need to be craeted with
+exclusive-lock and journaling image features. Additionally, This will also
+imply that NBD will need to be used instead of krbd.
+
+### Promotion of the rbd image
+
+During the **NodeStage request** from the CSI side, ceph-csi will promote the
+image and use it.
+
+If its a ROX or RWX PVC we need to handle image already mirrored case.
+
+Currently, **ceph-csi will not take care of demoting the rbd images, it will be
```

Comment:
I am still advocating for `ceph-csi` to automatically promote on `ControllerPublishVolume` and automatically demote on `ControllerUnpublishVolume`.